site/en/blog/wasmgc/index.md

```diff
@@ -0,0 +1,132 @@
+
+## Porting programming languages to new runtimes with WasmGC
+
+WasmGC is a [proposal](https://github.com/WebAssembly/gc/blob/main/proposals/gc/Overview.md) of the [WebAssembly Community Group](https://www.w3.org/community/webassembly/). The current Wasm MVP implementation is only capable of dealing with numbers, that is, integers and floats, in linear memory, and with the [reference types](https://github.com/WebAssembly/reference-types/blob/master/proposals/reference-types/Overview.md) proposal being shipped, Wasm can additionally hold on to external references. WasmGC now adds struct and array heap types, which means support for non-linear memory allocation. Each WasmGC object has a fixed type and structure, which makes it easy for VMs to generate efficient code to access their fields without the risk of [deoptimizations](https://web.dev/speed-v8/#de-optimization) that dynamic languages like JavaScript have. This proposal thereby adds efficient support for high-level managed languages to WebAssembly, via struct and array heap types that enable language compilers targeting Wasm to integrate with a garbage collector in the host VM. In simplified terms, this means that with WasmGC, porting a programming language to Wasm means the programming language's garbage collector no longer needs to be part of the port, but instead the existing garbage collector can be used.
+
+To verify the real-world impact of this improvement, Chrome's Wasm team has compiled versions of the [Fannkuch benchmark](https://benchmarksgame-team.pages.debian.net/benchmarksgame/description/fannkuchredux.html) (which allocates data structures as it works) from [C](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-gcc-5.html), [Rust](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-rust-2.html), and [Java](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-java-2.html). The C and Rust binaries could be anywhere from _6.1 K_ to _9.6 K_ depending on the various compiler flags, while the Java version is much smaller at only _2.3 K_! C and Rust do not include a garbage collector, but they do still bundle `malloc/free` to manage memory, and the reason Java is smaller here is because it doesn't need to bundle any memory management code at all. This is just one specific example, but it shows that WasmGC binaries have the potential of being very small, and this is even before any significant work on optimizing for size.
+
+## Seeing a WasmGC-ported programming language in action
+
+One of the first programming languages that has been ported to Wasm thanks to WasmGC is [Kotlin](https://kotlinlang.org/) in the form of [Kotlin/Wasm](https://kotl.in/wasmgc). The [demo](https://kotlin-wasm-hello-world.glitch.me/), with [source code](https://github.com/Kotlin/kotlin-wasm-examples/tree/main/browser-example) courtesy of the Kotlin team, is shown in the following listing.
```

Comment:
@tomayac please pull recent changes for examples and republish your examples before publishing the post.

I added more browser API usages to the simple browser example and made it a bit more interactive.
Also, other examples, including image viewer, got some cosmetic changes.

More importantly, all examples use the latest Kotlin toolchain, which generates final WasmGC opcodes.

Reply:
Done, both samples were updated to the latest version.

Reply:
Great, thanks!